File: android/app/src/main/java/com/openburnbar/data/media/MediaPartnerSavePreferenceStore.kt
package com.openburnbar.data.media

import android.content.Context
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

/**
 * 1:1 Kotlin port of `MediaPartnerSavePreferenceStore.swift`. Decision 3
 * of the Mercury Media master plan: per-paired-Mac save preferences for
 * inbound image attachments. First image from a given peer presents an
 * action sheet (Photos / Files); subsequent images use the saved choice.
 *
 * iOS persists in UserDefaults; Android uses DataStore Preferences keyed
 * by the peer iroh NodeId so the same per-partner persistence semantics
 * carry over and survive app updates.
 *
 * Settings → Media → "Per-partner save preferences" surfaces the list +
 * per-row "Forget" + global "Forget all".
 */
class MediaPartnerSavePreferenceStore(context: Context) {

    enum class SavePreference(val raw: String) {
        ASK_EACH_TIME("askEachTime"),
        PHOTOS("photos"),
        FILES("files");

        companion object {
            fun fromRaw(raw: String?): SavePreference? = values().firstOrNull { it.raw == raw }
        }
    }

    private val store = context.applicationContext.dataStore

    suspend fun preference(peerDeviceId: String): SavePreference {
        val raw = store.data.map { it[keyFor(peerDeviceId)] }.first()
        return SavePreference.fromRaw(raw) ?: SavePreference.ASK_EACH_TIME
    }

    fun preferenceFlow(peerDeviceId: String): Flow<SavePreference> =
        store.data.map { SavePreference.fromRaw(it[keyFor(peerDeviceId)]) ?: SavePreference.ASK_EACH_TIME }

    suspend fun setPreference(preference: SavePreference, peerDeviceId: String) {
        store.edit { prefs ->
            if (preference == SavePreference.ASK_EACH_TIME) {
                prefs.remove(keyFor(peerDeviceId))
            } else {
                prefs[keyFor(peerDeviceId)] = preference.raw
            }
        }
    }

    suspend fun forget(peerDeviceId: String) {
        store.edit { prefs -> prefs.remove(keyFor(peerDeviceId)) }
    }

    suspend fun forgetAll() {
        store.edit { prefs ->
            val keys = prefs.asMap().keys.filterIsInstance<Preferences.Key<String>>()
                .filter { it.name.startsWith(KEY_PREFIX) }
            keys.forEach { prefs.remove(it) }
        }
    }

    suspend fun storedPartners(): List<Pair<String, SavePreference>> {
        val snapshot = store.data.first()
        return snapshot.asMap().entries
            .mapNotNull { (key, value) ->
                if (!key.name.startsWith(KEY_PREFIX)) return@mapNotNull null
                val peerId = key.name.removePrefix(KEY_PREFIX)
                val pref = SavePreference.fromRaw(value as? String) ?: return@mapNotNull null
                peerId to pref
            }
            .sortedBy { it.first }
    }

    fun storedPartnersFlow(): Flow<List<Pair<String, SavePreference>>> = store.data.map { snapshot ->
        snapshot.asMap().entries
            .mapNotNull { (key, value) ->
                if (!key.name.startsWith(KEY_PREFIX)) return@mapNotNull null
                val peerId = key.name.removePrefix(KEY_PREFIX)
                val pref = SavePreference.fromRaw(value as? String) ?: return@mapNotNull null
                peerId to pref
            }
            .sortedBy { it.first }
    }

    private fun keyFor(peerDeviceId: String) = stringPreferencesKey(KEY_PREFIX + peerDeviceId)

    companion object {
        private const val KEY_PREFIX = "media.savePreference."
        private val Context.dataStore by preferencesDataStore(name = "mercury_media_partner_prefs")
    }
}

File: OpenBurnBarMobile/Services/Media/MediaPartnerSavePreferenceStore.swift
import Foundation

/// Implements Decision 3 of the Mercury media plan: per-paired-Mac save
/// preferences for inbound image attachments. First image from a given
/// peer presents an action sheet (Photos / Files); subsequent images
/// from the same peer use the saved choice. Persistence: `UserDefaults`
/// keyed by the peer iroh `NodeId`.
///
/// Settings → Media → "Per-partner save preferences" surfaces the list +
/// per-row "Forget" + global "Forget all".
public actor MediaPartnerSavePreferenceStore {
    public enum SavePreference: String, Sendable, Codable, Equatable {
        case askEachTime
        case photos
        case files
    }

    public static let shared = MediaPartnerSavePreferenceStore()

    private let defaults: UserDefaults
    private let keyPrefix: String

    public init(
        defaults: UserDefaults = .standard,
        keyPrefix: String = "media.savePreference."
    ) {
        self.defaults = defaults
        self.keyPrefix = keyPrefix
    }

    public func preference(forPeerDeviceId peerDeviceId: String) -> SavePreference {
        let key = keyPrefix + peerDeviceId
        guard let raw = defaults.string(forKey: key),
              let value = SavePreference(rawValue: raw) else {
            return .askEachTime
        }
        return value
    }

    public func setPreference(
        _ preference: SavePreference,
        forPeerDeviceId peerDeviceId: String
    ) {
        let key = keyPrefix + peerDeviceId
        if preference == .askEachTime {
            defaults.removeObject(forKey: key)
        } else {
            defaults.set(preference.rawValue, forKey: key)
        }
    }

    public func forget(peerDeviceId: String) {
        defaults.removeObject(forKey: keyPrefix + peerDeviceId)
    }

    public func forgetAll() {
        let domain = defaults.dictionaryRepresentation()
        for key in domain.keys where key.hasPrefix(keyPrefix) {
            defaults.removeObject(forKey: key)
        }
    }

    /// Snapshot of every paired Mac with a stored preference. Powers the
    /// Settings → Media UI.
    public func storedPartners() -> [(peerDeviceId: String, preference: SavePreference)] {
        let domain = defaults.dictionaryRepresentation()
        var result: [(String, SavePreference)] = []
        for (key, value) in domain {
            guard key.hasPrefix(keyPrefix) else { continue }
            let id = String(key.dropFirst(keyPrefix.count))
            if let raw = value as? String, let pref = SavePreference(rawValue: raw) {
                result.append((id, pref))
            }
        }
        return result.sorted { $0.0 < $1.0 }
    }
}
